### DISCUSSION TITLE
How to keep Expose Server alive on Windows

### DISCUSSION THREAD

**Discussion by bastianhilton** [exposedev/expose#227] [Q&A]
I have read the documentation here: https://beyondco.de/docs/expose/server/starting-the-server

How to keep it alive via Mac OS and Linux but there aren't ways to keep it alive on Windows. Normally you could easily do it if Expose was an application or a some sort of file, but since its commands i'm curious of how to do that on windows.

The current documentation works for windows except the part to keep Expose alive just in case the command prompt is exited.

Any help would be greatly appreciated.

**Comment by mpociot** [comment]
Sorry, I don't have any experience in hosting your own Expose server on Windows.
If you (or anyone else) has any experience with this and can help improve the documentation, I would appreciate a PR.

**Comment by plabbett** [comment]
I was able to modify [this Powershell script](https://raw.githubusercontent.com/JFLarvoire/SysToolsLib/master/PowerShell/PSService.ps1
) that creates a Windows Service.  I installed Expose within the project and use the Powershell script/service to check if expose is running and start it if it's not. 

The timer tick looks similar to below. 
```ps1
     $running = gcim win32_process | where-object {$_.Name -like "*php.exe*"} | select commandline | where-object {$_.commandline -like "*expose*share-port 1433"}

           if($running){
            Log "$scriptName -Service # Tunnel Is running"
           }
           else
           {
               #$installDir = "${ENV:ProgramFiles}\$serviceName"
               $cmd = "'$($installDir)\vendor\bin\expose' share-port 1433"
               Log "$scriptName -Service # Attempting to restart tunnel"
               $cmd = "$($installDir)\vendor\bin\expose"
               Start-Process -FilePath $cmd -WorkingDirectory $installDir -ArgumentList "share-port 1433"

               $cmd = $null;
           }
           $running = $null;
```

My project directory looks like this (I renamed the file to `CallTheoryDashboardTunnel` in the screenshot)

![image](https://user-images.githubusercontent.com/1407028/123533286-00d76080-d6e2-11eb-9b57-1dfcd0f2fc68.png)

- Expose available through composer at `vendor/bin/expose` in per-project installation
- Modified config `.expose.php` to load token from `api_key.txt` and use custom servers
- Use the powershell script to create the service
- Use Windows Service MMC to manage the service
- Composer `post-install-cmd` creates the Log folder

A complete install looks like this:

```
choco install php composer
# Extract contents to C:\Program Files\ServiceName\
cd "C:\Program Files\ServiceNameFolder"
composer install
.\ServiceName.ps1 -Setup
.\ServiceName.ps1 -Start
```

I'm only using this in scenarios where I have no other option but to run via Windows.  Hope it helps!
